File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> Genres, vector<int> Plays) {
    using GenreInfo = pair<int, vector<pair<int, int>>>;
    
    unordered_map<string, GenreInfo> SongInfoByGenre;
    for(int i = 0; i < Genres.size(); ++i)
    {
        const string& Genre = Genres[i]; const int Play = Plays[i];
        auto& [TotalPlayTime, SongList] = SongInfoByGenre[Genre];
        
        TotalPlayTime += Play;
        SongList.emplace_back(i, Play);
    }
    
    vector<GenreInfo> ListOfGenreInfo(SongInfoByGenre.size());
    transform(SongInfoByGenre.begin(), SongInfoByGenre.end(), ListOfGenreInfo.begin(), [](auto pair) { return pair.second; });
    
    sort(ListOfGenreInfo.begin(), ListOfGenreInfo.end(), greater<>());
    for(auto& [TotalPlayTime, SongList] : ListOfGenreInfo)
    {
        sort(SongList.begin(), SongList.end(), [](auto p, auto q) 
             {
                 if(p.second == q.second)
                 {
                     return p.first < q.first;
                 }
                 return p.second > q.second;
             });
    }
    
    vector<int> BestAlbum;
    for(const auto& [TotalPlayTime, SongList] : ListOfGenreInfo)
    {
        int SelectCnt = 0;
        for(const auto& [Id, Play] : SongList)
        {
            if(SelectCnt == 2)
            {
                break;
            }
            
            BestAlbum.emplace_back(Id);
            SelectCnt += 1;
        }
    }
    
    return BestAlbum;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    playlist_by_genre = {genre : [] for genre in set(genres)}
    for i, (genre, play) in enumerate(zip(genres, plays)):
        playlist_by_genre[genre].append((play, i))
                                      
    sorted_genre = sorted(playlist_by_genre.keys(), key=lambda x: sum(map(lambda y: y[0], playlist_by_genre[x])), reverse = True)
    
    answer = []
    for genre in sorted_genre:
        temp = [index for play, index in sorted(playlist_by_genre[genre], key=lambda x: (x[0], -x[1]), reverse=True)]
        answer += temp[:2]
    return answer